modules/javafx.graphics/src/main/java/javafx/scene/image/PixelBuffer.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.scene.image;
+
+import com.sun.javafx.geom.Rectangle;
+import com.sun.javafx.tk.Toolkit;
+import javafx.geometry.Rectangle2D;
+import javafx.util.Callback;
+
+import java.lang.ref.WeakReference;
+import java.nio.Buffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * A class to provide a shared pixel data using {@link java.nio.Buffer} for {@code WritableImage}s.
+ * {@code PixelBuffer} supports only {@code PixelFormat.INT_ARGB_PRE}
+ * and {@code PixelFormat.BYTE_BGRA_PRE} pixel formats.
+ * A {@code PixelBuffer} can only be drawn using a {@code WritableImage}.
+ * Multiple {@code WritableImage}s and applications can share a {@code PixelBuffer}.
+ * <p>
+ * Pseudo code to create {@code PixelBuffer}:
+ * <pre>{@code
+ *
+ * // Creating a PixelBuffer using PixelFormat.BYTE_BGRA_PRE
+ * ByteBuffer byteBuffer = ByteBuffer.allocateDirect(width * height * 4);
+ * PixelFormat<ByteBuffer> pixelFormat = PixelFormat.getByteBgraPreInstance();
+ * PixelBuffer pixelBuffer = new PixelBuffer(width, height, byteBuffer, pixelFormat);
+ * Image img = new WritableImage(pixelBuffer);
+ *
+ * // Creating a PixelBuffer using PixelFormat.INT_ARGB_PRE
+ * IntBuffer intBuffer = IntBuffer.allocate(width * height);
+ * PixelFormat<IntBuffer> pixelFormat = PixelFormat.getIntArgbPreInstance();
+ * PixelBuffer pixelBuffer = new PixelBuffer(width, height, intBuffer, pixelFormat);
+ * Image img = new WritableImage(pixelBuffer);
+ * }</pre>
+ *
+ * @since 13
+ */
+public class PixelBuffer<T extends Buffer> {
+
+    private final T buffer;
+    private final int width;
+    private final int height;
+    private final PixelFormat pixelFormat;
```

Comment:
`PixelBuffer<ByteBuffer>` can have  `PixelFormat<IntBuffer>`  when the `PixelBuffer `is created using raw type call,
`PixelBuffer pb = new PixelBuffer(w, h, IntBuffer.allocate(w * h), PixelFormat.getByteBgraPreInstance());`
I think a strict type check should be added to PixelBuffer constructor to avoid this scenario:
        ```
if ((pixelFormat.getType() == PixelFormat.Type.BYTE_BGRA_PRE
                 && !(buffer instanceof ByteBuffer)) ||
                (pixelFormat.getType() == PixelFormat.Type.INT_ARGB_PRE
                     && !(buffer instanceof IntBuffer))) {
            // May occur when the constructor is called in raw type(unchecked conversion warning)
            throw new IllegalArgumentException("Incorrect combination of PixelFormat and Buffer.");
        }
```

Reply:
When I wrote " I don't think I can have..." I meant "I don't think I *should* have...". Since you agree with this, by declaring `PixelFormat<T>` you add a compile time check to the constructor that they both use the same buffer type. All constructor checks are only runtime. If you use `PixelBuffer(..., T buffer, PixelFormat<T> pixelFormat)`, the `instanceof` check is not needed.
This also makes `PixelFormat<T> getPixelFormat()` return the correct type.

In general, there's very little reason to use raw types.

---

As an aside, the "problem" with the way `PixelFormat` is designed is that the buffer type does not imply a single format type (but it does restrict the formats). The opposite is true though - a format type implies a single buffer type. If `PixelFormat` was coded that way (and I'm not sure it could have because there are no generic enums yet), this would have been much simpler.